nixos/modules/services/backup/duplicity.nix

```diff
@@ -0,0 +1,146 @@
+        List of paths to exclude from backups. See
+        <citerefentry><refentrytitle>duplicity</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry> for details on the syntax.
+      '';
+    };
+
+    targetUrl = mkOption {
+      type = types.str;
+      example = "s3://host:port/prefix";
+      description = ''
+        Target url to backup to. See
+        <citerefentry><refentrytitle>duplicity</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry> for supported urls.
+      '';
+    };
+
+    secretFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        Path of a file containing secrets (gpg passphrase, access key...) in
+        the format of EnvironmentFile as described by
+        <citerefentry><refentrytitle>systemd.exec</refentrytitle>
+        <manvolnum>5</manvolnum></citerefentry>. For example:
+        <programlisting>
+        PASSPHRASE=<replaceable>...</replaceable>
+        AWS_ACCESS_KEY_ID=<replaceable>...</replaceable>
+        AWS_SECRET_ACCESS_KEY=<replaceable>...</replaceable>
+        </programlisting>
+      '';
+    };
+
+    frequency = mkOption {
+      type = types.nullOr types.str;
+      default = "daily";
+      description = ''
+        Run duplicity with the given frequency (see
+        <citerefentry><refentrytitle>systemd.time</refentrytitle>
+        <manvolnum>7</manvolnum></citerefentry> for the format).
+        If null, do not run automatically.
+      '';
+    };
+
+    extraFlags = mkOption {
+      type = types.listOf types.str;
+      default = [];
+      example = [ "--full-if-older-than" "1M" ];
+      description = ''
+        Extra command-line flags passed to duplicity. See
+        <citerefentry><refentrytitle>duplicity</refentrytitle>
+        <manvolnum>1</manvolnum></citerefentry>.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.duplicity = {
+      description = "backup files with duplicity";
+
+      script = ''
+        exec ${pkgs.duplicity}/bin/duplicity ${escapeShellArgs (
+          [
+            cfg.root
+            cfg.targetUrl
+            "--archive-dir" stateDirectory
+          ]
+          ++ concatMap (p: [ "--include" p ]) cfg.include
+          ++ concatMap (p: [ "--exclude" p ]) cfg.exclude
+          ++ cfg.extraFlags)}
+      '';
+
+      environment = {
+        GNUPGHOME=stateDirectory;
+      };
+
+      serviceConfig = {
+        PrivateTmp = true;
+        ProtectSystem = "strict";
+        ProtectHome = "read-only";
+        StateDirectory = builtins.baseNameOf stateDirectory;
+      } // optionalAttrs hasLocalTarget {
+        ReadWritePaths = localTarget;
+      } // optionalAttrs (cfg.secretFile != null) {
+        EnvironmentFile = cfg.secretFile;
+      };
+
+    } // optionalAttrs (cfg.frequency != null) {
+      startAt = cfg.frequency;
+    };
+
+    system.activationScripts.duplicity = ''
+      mkdir -m 0700 -p ${escapeShellArg localTarget}
```

Comment:
You can use the `systemd.tmpfiles` option instead to create this directory declaratively (look through nixpkgs for examples). And this should be wrapped with a check whether it's even a local target (see first comment as well).